xonsh/history.py

```diff
@@ -407,103 +406,84 @@ def _hist_create_parser():
-    which will return the history as a list with each item
-    in the tuple form (name, start_time, index).
+def _hist_get_session(session='session', location=None):
+    "Yield commands from history."
+    cmds = _HIST_SESSIONS[session](location=location)
+    try:
+        yield from cmds
+    except TypeError:
+        pass
+
+
+def _hist_get_portion(commands, slices):
+    "Yield from portions of history commands. "
+    commands = list(commands)
+    for s in slices:
+        s = ensure_slice(s)
+        yield from commands[s]
+
+
+def _hist_filter_ts(commands, start_time=None, end_time=None):
+    """Yield only the commands between start and end time."""
+    if start_time is None:
+        start_time = 0.0
+    if end_time is None:
+        end_time = float('inf')
+    for cmd in commands:
+        if end_time > cmd[1] > start_time:
+            yield cmd
+
+
+def _hist_get(session='session', slices=None,
+              start_time=None, end_time=None, location=None):
+    """Get the requested portion of shell history.
+
+    Parameters
+    ----------
+    session: {'session', 'all', 'xonsh', 'bash', 'zsh'}
+        The history session to get.
+    slices : list of slice-like objects, optional
+        The portions of history to yield.
+    start_time, end_time: float, optional
+        Filter commands by timestamp.
+    location: string, optional
+        The history file location (bash or zsh)
+
+    Returns
+    -------
+    list
+       A filtered list of commands
+    """
+    cmds = _hist_get_session(session, location)
+    if slices:
+        cmds = _hist_get_portion(cmds, slices)
+    if start_time or end_time:
+        cmds = _hist_filter_ts(cmds, start_time, end_time)
+    return list(cmds)
 
-    If invoked via __xonsh_history__.show() then the ns parameter
-    can be supplied as a str with the follow options::
 
-        session - returns xonsh history from current session
-        xonsh   - returns xonsh history from all sessions
-        all     - alias of xonsh
-        zsh     - returns all zsh history
-        bash    - returns all bash history
-    """
-    # Check if ns is a string, meaning it was invoked from
-    if hist is None:
-        hist = bultins.__xonsh_history__
-    alias = True
-    if isinstance(ns, str) and ns in _HIST_SESSIONS:
-        ns = _hist_create_parser().parse_args([ns])
-        alias = False
-    if not ns:
-        ns = _hist_create_parser().parse_args(['show', 'xonsh'])
-        alias = False
+def _hist_show(ns, *args, **kwargs):
+    """Show the requested portion of shell history. Accepts same parameters
+    with `_hist_get`."""
     try:
-        commands = _HIST_SESSIONS[ns.session](hist=hist, location=location)
-    except KeyError:
-        print("{} is not a valid history session".format(ns.action))
-        return None
+        commands = _hist_get(ns.session, ns.slices, **kwargs)
+    except ValueError as err:
+        print(err, file=sys.stderr)
```

Comment:
I think the previous error message was a little more informative.  Maybe you could merge the two.